[pipelining] Add ZBV schedule

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142084

Adds ZBV schedule which is explained in https://arxiv.org/pdf/2401.10241, Section 6. Tested it works under the new PipelineScheduleRuntime by fixing a small bug in handling V-shaped schedules. This PR is a replacement for https://github.com/pytorch/pytorch/pull/138444

cc the original authors: @QPHutu @ufotalent https://github.com/pytorch/pytorch/pull/138444#issuecomment-2472684977

cc @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

H-Huang committed:
[pipelining] Add ZBV schedule

[ghstack-poisoned]

H-Huang committed:
Update on "[pipelining] Add ZBV schedule"


Adds ZBV schedule which is explained in https://arxiv.org/pdf/2401.10241, Section 6. Tested it works under the new PipelineScheduleRuntime by fixing a small bug in handling V-shaped schedules.

cc the original authors: QPHutu ufotalent

cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[pipelining] Add ZBV schedule"


Adds ZBV schedule which is explained in https://arxiv.org/pdf/2401.10241, Section 6. Tested it works under the new PipelineScheduleRuntime by fixing a small bug in handling V-shaped schedules.

cc the original authors: QPHutu ufotalent

cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -2160,6 +2161,165 @@ def need_bubble(stage, op, microbatch, num_stages_global, seen_ops):
+            rank_ops = self._calculate_single_rank_operations(rank)
+            self.pipeline_order[rank] = rank_ops
+
+    def _calculate_single_rank_operations(self, rank) -> List[Optional[_Action]]:
```
nit: maybe consider adding more comments (esp for variables) in this function?

kwen2501 commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -2160,6 +2161,165 @@ def need_bubble(stage, op, microbatch, num_stages_global, seen_ops):
+
+        assert w0_cnt == b0_cnt and b0_cnt == f0_cnt
+        assert w1_cnt == b1_cnt and b1_cnt == f1_cnt
+        # remove redundant microbatches
+        rank_ops = [
+            (
+                action
+                if action is not None
+                and action.microbatch_index is not None
+                and action.microbatch_index < self._n_microbatches
+                else None
+            )
+            for action in rank_ops
+        ]
```
what this step for?

H-Huang replied:
Added a comment, the beginning of the function sets the `n_microbatch` count to a value in order to fully fill the pipeline. However, if the actual number of microbatches (passed by user) is less than desired than these are just removed rather than erroring out.

kwen2501 reviewed (approved):
Nice work! The addition of a schedule looks pretty easy now.

wconstab commented on test/distributed/pipelining/test_schedule_multiproc.py:
```diff
@@ -710,9 +711,20 @@ def test_pipeline_schedule_runtime_custom_sched(self, ScheduleClass):
+        "ScheduleClass", [ScheduleVShaped, ScheduleUnbalanced, ScheduleZBVZeroBubble]
+    )
+    @parametrize("use_new_runtime", [False, True])
+    def test_non_symmetric_stage_ids(self, ScheduleClass, use_new_runtime):
```
nit: capitalized ScheduleClass ends up looking kinda weird here.  maybe `schedule_class`?

wconstab commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -2160,6 +2161,165 @@ def need_bubble(stage, op, microbatch, num_stages_global, seen_ops):
+    """
+    The Zero Bubble schedule (ZBV variant).
+    See https://arxiv.org/pdf/2401.10241 Section 6 for details.
+    This schedule wille perform one forward and one backward on inputs for the microbatches in steady
```
typo wille

wconstab commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -2160,6 +2161,165 @@ def need_bubble(stage, op, microbatch, num_stages_global, seen_ops):
+    The Zero Bubble schedule (ZBV variant).
+    See https://arxiv.org/pdf/2401.10241 Section 6 for details.
+    This schedule wille perform one forward and one backward on inputs for the microbatches in steady
+    state and supports multiple stages per rank. Uses the backward for weights to fill in
```
does this schedule actually require 2 stages per rank? or does it support more than 2 stages per rank

wconstab replied:
it would also be good to mention that this ZB-V schedule would have the "zero bubble" property only if tF=tI=tW, which is likely not strictly true for real models.  Alternative greedy scheduler could be implemented for unequal/unbalanced time

H-Huang committed:
Update on "[pipelining] Add ZBV schedule"


Adds ZBV schedule which is explained in https://arxiv.org/pdf/2401.10241, Section 6. Tested it works under the new PipelineScheduleRuntime by fixing a small bug in handling V-shaped schedules. This PR is a replacement for https://github.com/pytorch/pytorch/pull/138444

cc the original authors: QPHutu ufotalent https://github.com/pytorch/pytorch/pull/138444#issuecomment-2472684977

cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge
